### PR TITLE
Refactor subflow redirect

### PR DESCRIPTION
**Pull request by rafael-santiago-take**
Refactored the way of occour the redirection to subflow

### CODE REVIEW

File: src/Take.Blip.Builder/FlowManager.cs [hunk 1]
```diff
@@ -352,6 +375,55 @@ public async Task ProcessInputAsync(Message message, Flow flow, IContext message
...
+            var newState = subflow.States.Single(s => s.Root);
+
+            await _stateSessionManager.SetStateAsync(userIdentity, shortNameOfSubflow, cancellationToken);
+
+            return (subflow, newState);
+        }
+
+        private async Task<(Flow, State)> RedirectToParentFlowAsync(IContext context, Identity userIdentity, Flow flow, string parentStateId, CancellationToken cancellationToken)
+        {
+            if (context.Flow.Parent == null)
```
Comment (caiopsouza) [inline src/Take.Blip.Builder/FlowManager.cs:411]: You check if `context.Flow.Parent` is null but this value is never read.

File: src/Take.Blip.Builder/StateManager.cs [hunk 2]
```diff
@@ -22,6 +22,11 @@ public Task<string> GetPreviousStateIdAsync(IContext context, CancellationToken
             return context.GetContextVariableAsync(GetPreviousStateKey(context.Flow.Id), cancellationToken);
         }
 
+        public Task<string> GetParentStateIdAsync(IContext context, CancellationToken cancellationToken)
+        {
+            return context.GetContextVariableAsync(GetStateKey(context.Flow.Parent.Id), cancellationToken);
```
Comment (caiopsouza) [inline src/Take.Blip.Builder/StateManager.cs:27]: I saw somewhere else checking if `context.Flow.Parent` is null. Could it also be null here too?

File: src/Take.Blip.Builder/FlowManager.cs [hunk 3]
```diff
@@ -352,6 +375,55 @@ public async Task ProcessInputAsync(Message message, Flow flow, IContext message
...
+
+            context.Flow = parentFlow;
+            var state = parentFlow.States.FirstOrDefault(s => s.Id == parentStateId) ?? parentFlow.States.Single(s => s.Root);
+
+            await _stateSessionManager.SetStateAsync(userIdentity, parentFlow.SessionState, cancellationToken);
+
+            return (parentFlow, state);
+        }
+
+        private bool isSubflowState(State state) => state != null && state.Id.StartsWith("subflow:");
```
[the remark this answers is not in the document]
Comment (rafael-santiago-take) [reply src/Take.Blip.Builder/FlowManager.cs]: done

File: src/Take.Blip.Builder/StateManager.cs (hunk 2 above)
Comment (rafael-santiago-take) [reply src/Take.Blip.Builder/StateManager.cs:32]: added validation

File: src/Take.Blip.Builder/FlowManager.cs (hunk 1 above)
Comment (rafael-santiago-take) [reply src/Take.Blip.Builder/FlowManager.cs]: fixed

File: src/Take.Blip.Builder/Models/Flow.cs [hunk 4]
```diff
@@ -14,6 +14,9 @@ namespace Take.Blip.Builder.Models
     /// </summary>
     public class Flow : IValidable
     {
+        private const string DEFAULT_SESSION_STATE = "default";
+        private const int MINIMUM_SUBFLOW_VERSION = 2;
```
[the remark this answers is not in the document]
Comment (rafael-santiago-take) [reply src/Take.Blip.Builder/Models/Flow.cs]: done

File: src/Take.Blip.Builder/FlowManager.cs [hunk 5]
```diff
@@ -352,6 +375,56 @@ public async Task ProcessInputAsync(Message message, Flow flow, IContext message
             }
         }
 
+        private async Task ProcessGlobalOutputActions(IContext context, Flow flow, LazyInput lazyInput, InputTrace inputTrace, CancellationToken cancellationToken)
```
Comment (jujucasseb) [inline src/Take.Blip.Builder/FlowManager.cs:378]: add Async sufix

File: src/Take.Blip.Builder/FlowManager.cs [hunk 6]
```diff
@@ -545,5 +618,17 @@ await output.Conditions.EvaluateConditionsAsync(lazyInput, context, cancellation
 
             return state;
         }
+
+        private async Task<string> GetParentStateId(IContext context, Queue<string> parentStateIdQueue, CancellationToken cancellationToken) => parentStateIdQueue.Count > 0 ? parentStateIdQueue.Dequeue() : await _stateManager.GetParentStateIdAsync(context, cancellationToken);
```
Comment (jujucasseb) [inline src/Take.Blip.Builder/FlowManager.cs:622]: add Async sufix

Comment (rafael-santiago-take) [reply src/Take.Blip.Builder/FlowManager.cs]: done

**Pull request merged**